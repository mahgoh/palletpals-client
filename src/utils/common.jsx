export function classNames(...classes) {
  return classes.filter(Boolean).join(' ')
}

export function productImageURL(imageId) {
  return `${GLOBAL.API_URL}/product-images/${imageId}`
}

export function parseDateTime(dateTimeString, locale = 'en') {
  const datetime = new Date(dateTimeString)
  const options = {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: 'numeric',
    minute: '2-digit',
  }

  return datetime.toLocaleString(locale, options)
}

export function formatPrice(price) {
  return `CHF ${price.toFixed(2)}`
}

export function formatOrderId(id) {
  return id.toString().padStart(6, '0')
}
