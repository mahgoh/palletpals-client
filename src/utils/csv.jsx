export async function processCSV(file) {
  try {
    const csv = await readCSV(file)
    const parsed = parseCSV(csv)

    return parsed
  } catch (e) {
    throw e
  }
}

function readCSV(file) {
  return new Promise((resolve, reject) => {
    if (file === null) reject('No file selected')

    // Check if the file is an image.
    if (file.type && !file.type === 'text/csv') {
      console.log('File is not csv.', file.type, file)
      reject('File is not csv.')
    }

    const reader = new FileReader()
    reader.addEventListener('load', (event) => {
      resolve(event.target.result)
    })
    reader.readAsText(file)
  })
}

/**
 *
 * Target:
 * {
 *   "km": [10,20,30],
 *   "pallets": [1,2,3],
 *   "price": [[5,6,7],[9,11,13],[15,17,19]]
 * }
 * @param {string} csv
 * @returns
 */

function parseCSV(csv) {
  const lines = csv.split('\n')

  // how entries are separated
  const delimiter = csv.includes(',') ? ',' : ';'

  const head = lines.shift()

  // palletArray is the first row except the first column
  // integers
  const palletArray = head
    .split(delimiter)
    .slice(1)
    .map((p) => parseInt(p))

  // kmArray is the first column except the first row
  // integers
  const kmArray = lines.map(
    (line) => line.split(delimiter).map((km) => parseInt(km))[0]
  )

  // priceMatrix is the rest of the table as two dimensional array
  // floats
  const priceMatrix = lines.map((line) =>
    line
      .split(delimiter)
      .slice(1)
      .map((p) => parseFloat(p))
  )

  return {
    kmArray,
    palletArray,
    priceMatrix,
  }
}
