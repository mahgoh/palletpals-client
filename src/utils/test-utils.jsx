import { cloneElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { I18nextProvider } from 'react-i18next'
import { render } from '@testing-library/react'
import i18n from '../i18n'

import { AuthProvider } from '@/services/auth'
import { AppearanceProvider } from '@/services/appearance'
import { CartProvider } from '@/services/cart'
import { LanguageProvider } from '@/services/language'
import { NotificationProvider } from '@/services/notification'

/**
 * Render component with router wrapper and option for specifying the current route.
 *
 * @param {JSX} component
 * @param {Object} options
 * @returns
 */
const renderWithRouter = (component, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route)

  return {
    ...render(component, { wrapper: BrowserRouter }),
  }
}

/**
 * Render component with router wrapper and i18next provider.
 * @param {JSX} component
 * @param {Object} options
 * @returns
 */
const renderWithRouterWithi18n = (component, options) => {
  const comp = cloneElement(component, {
    changeLanguage: (lng) => {
      i18n.changeLanguage(lng)
      rerender(<I18nextProvider i18n={i18n}>{comp}</I18nextProvider>)
    },
  })
  const defaultRender = renderWithRouter(
    <I18nextProvider i18n={i18n}>{comp}</I18nextProvider>,
    options
  )
  const { rerender } = defaultRender
  return defaultRender
}

const renderComplete = (component, options) => {
  return {
    ...renderWithRouterWithi18n(
      <AuthProvider>
        <AppearanceProvider>
          <LanguageProvider>
            <CartProvider>
              <NotificationProvider>{component}</NotificationProvider>
            </CartProvider>
          </LanguageProvider>
        </AppearanceProvider>
      </AuthProvider>,
      options
    ),
  }
}

/**
 * Render component with specified renderer
 *
 * @param {JSX} component
 * @param {Object} param1
 * @returns
 */
const customRender = (component, { renderer = 'default', ...args } = {}) => {
  switch (renderer) {
    case 'router':
      return renderWithRouter(component, { ...args })
    case 'router-i18n':
      return renderWithRouterWithi18n(component, { ...args })
    case 'complete':
      return renderComplete(component, { ...args })
    case 'default':
    default:
      return {
        ...render(component),
      }
  }
}

export * from '@testing-library/react'
// export { default as userEvent } from '@testing-library/user-event'
export { customRender as render }
