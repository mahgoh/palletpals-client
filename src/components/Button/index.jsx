import PropTypes from 'prop-types'
import { useNavigate } from 'react-router-dom'
import { classNames } from '@/utils/common'

export function LinkButton({ to, ...props }) {
  const navigate = useNavigate()

  if (!to) {
    return null
  }

  function onClick(e) {
    e.preventDefault()
    navigate(to)
  }

  return <Button onClick={onClick} {...props}></Button>
}

LinkButton.propTypes = {
  to: PropTypes.string.isRequired,
}

export default function Button({
  children,
  onClick,
  className,
  color = 'primary',
  ...props
}) {
  const colorClasses = {
    primary:
      'bg-orange-500 focus:bg-orange-700 hover:bg-orange-700 focus:ring-orange-500 text-white',
    secondary:
      'bg-orange-100 focus:bg-orange-200 hover:bg-orange-200 focus:ring-orange-500 text-orange-500',
    red: 'bg-rose-500 focus:bg-rose-700 hover:bg-rose-700 focus:ring-rose-500 text-white',
    redOutline:
      'bg-rose-500/20 focus:bg-rose-500 hover:bg-rose-500 focus:ring-rose-500 text-rose-500 hover:text-white focus:text-white',
  }

  return (
    <button
      className={classNames(
        className,
        colorClasses[color],
        'rounded-md border border-transparent py-3 px-8 text-base font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2'
      )}
      onClick={onClick}
      {...props}
    >
      {children}
    </button>
  )
}

Button.propTypes = {
  onClick: PropTypes.func,
  className: PropTypes.string,
  color: PropTypes.oneOf(['primary', 'secondary', 'red', 'redOutline']),
}
