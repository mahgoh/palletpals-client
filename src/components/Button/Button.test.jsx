import React from 'react'
import { vi } from 'vitest'
import { render, screen } from '@/utils/test-utils'
import Button, { LinkButton } from '.'

describe('button', () => {
  it('should render', () => {
    render(<Button>Click</Button>)
    expect(screen.getByRole('button')).toHaveTextContent('Click')
  })

  it('should render additional classNames', () => {
    render(<Button className="absolute">Click</Button>)
    expect(screen.getByRole('button')).toHaveTextContent('Click')
    expect(screen.getByRole('button')).toMatchSnapshot()
  })

  it('should render primary button', () => {
    render(<Button color="primary">Click</Button>)
    expect(screen.getByRole('button')).toMatchSnapshot()
  })

  it('should render secondary button', () => {
    render(<Button color="secondary">Click</Button>)
    expect(screen.getByRole('button')).toMatchSnapshot()
  })

  it('should render red button', () => {
    render(<Button color="red">Click</Button>)
    expect(screen.getByRole('button')).toMatchSnapshot()
  })

  it('should handle onClick', () => {
    const onClick = vi.fn()
    render(<Button onClick={onClick}>Click</Button>)

    expect(screen.getByRole('button')).toHaveTextContent('Click')
    expect(onClick).not.toHaveBeenCalled()

    screen.getByRole('button').click()
    expect(onClick).toHaveBeenCalled()
  })
})

describe('link button', () => {
  it('should render', () => {
    render(<LinkButton to="/page">Click</LinkButton>, { renderer: 'router' })
    expect(screen.getByRole('button')).toHaveTextContent('Click')
  })

  it('should not render', () => {
    const { container } = render(<LinkButton>Click</LinkButton>, {
      renderer: 'router',
    })
    expect(container).toBeEmptyDOMElement()
  })

  it('should render additional classNames', () => {
    render(
      <LinkButton to="/page" className="absolute">
        Click
      </LinkButton>,
      { renderer: 'router' }
    )
    expect(screen.getByRole('button')).toHaveTextContent('Click')
    expect(screen.getByRole('button')).toMatchSnapshot()
  })

  it('should render primary button', () => {
    render(
      <LinkButton to="/page" color="primary">
        Click
      </LinkButton>,
      { renderer: 'router' }
    )
    expect(screen.getByRole('button')).toMatchSnapshot()
  })

  it('should render secondary button', () => {
    render(
      <LinkButton to="/page" color="secondary">
        Click
      </LinkButton>,
      { renderer: 'router' }
    )
    expect(screen.getByRole('button')).toMatchSnapshot()
  })

  it('should render red button', () => {
    render(
      <LinkButton to="/page" color="red">
        Click
      </LinkButton>,
      { renderer: 'router' }
    )
    expect(screen.getByRole('button')).toMatchSnapshot()
  })
})
