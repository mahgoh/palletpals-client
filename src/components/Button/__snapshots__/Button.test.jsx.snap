// Vitest Snapshot v1

exports[`button > should render additional classNames 1`] = `
<button
  class="absolute bg-orange-500 focus:bg-orange-700 hover:bg-orange-700 focus:ring-orange-500 text-white px-6 py-2 sm:py-3 sm:px-8 rounded-md border border-transparent  text-base font-medium ring-offset-transparent transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 "
>
  Click
</button>
`;

exports[`button > should render primary button 1`] = `
<button
  class="bg-orange-500 focus:bg-orange-700 hover:bg-orange-700 focus:ring-orange-500 text-white px-6 py-2 sm:py-3 sm:px-8 rounded-md border border-transparent  text-base font-medium ring-offset-transparent transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 "
>
  Click
</button>
`;

exports[`button > should render red button 1`] = `
<button
  class="bg-rose-500 focus:bg-rose-700 hover:bg-rose-700 focus:ring-rose-500 text-white px-6 py-2 sm:py-3 sm:px-8 rounded-md border border-transparent  text-base font-medium ring-offset-transparent transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 "
>
  Click
</button>
`;

exports[`button > should render red outline button 1`] = `
<button
  class="bg-rose-500/20 focus:bg-rose-500 hover:bg-rose-500 focus:ring-rose-500 text-rose-500 hover:text-white focus:text-white px-6 py-2 sm:py-3 sm:px-8 rounded-md border border-transparent  text-base font-medium ring-offset-transparent transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 "
>
  Click
</button>
`;

exports[`button > should render secondary button 1`] = `
<button
  class="bg-orange-100 focus:bg-orange-200 hover:bg-orange-200 focus:ring-orange-500 text-orange-500 px-6 py-2 sm:py-3 sm:px-8 rounded-md border border-transparent  text-base font-medium ring-offset-transparent transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 "
>
  Click
</button>
`;

exports[`link button > should render additional classNames 1`] = `
<button
  class="absolute bg-orange-500 focus:bg-orange-700 hover:bg-orange-700 focus:ring-orange-500 text-white px-6 py-2 sm:py-3 sm:px-8 rounded-md border border-transparent  text-base font-medium ring-offset-transparent transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 "
>
  Click
</button>
`;

exports[`link button > should render primary button 1`] = `
<button
  class="bg-orange-500 focus:bg-orange-700 hover:bg-orange-700 focus:ring-orange-500 text-white px-6 py-2 sm:py-3 sm:px-8 rounded-md border border-transparent  text-base font-medium ring-offset-transparent transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 "
>
  Click
</button>
`;

exports[`link button > should render red button 1`] = `
<button
  class="bg-rose-500 focus:bg-rose-700 hover:bg-rose-700 focus:ring-rose-500 text-white px-6 py-2 sm:py-3 sm:px-8 rounded-md border border-transparent  text-base font-medium ring-offset-transparent transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 "
>
  Click
</button>
`;

exports[`link button > should render secondary button 1`] = `
<button
  class="bg-orange-100 focus:bg-orange-200 hover:bg-orange-200 focus:ring-orange-500 text-orange-500 px-6 py-2 sm:py-3 sm:px-8 rounded-md border border-transparent  text-base font-medium ring-offset-transparent transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 "
>
  Click
</button>
`;
